### PR TITLE
All nodes request helpers

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**

### CODE REVIEW

**Comment by gavin-norman-sociomantic** [comment]
Based on #19 (based on #8). Blocked until parent PRs are merged.

**Comment by gavin-norman-sociomantic** [comment]
Needs release notes.

**Comment by gavin-norman-sociomantic** [comment]
Rebased on v4.x.x with #19 merged. Unblocked.

**Comment by nemanja-boric-sociomantic** [comment]
Redish

**Comment by gavin-norman-sociomantic** [comment]
Ha, that curious D1 error where you can shove template parameters on something that's not a template -- and the compiler ignores them!

**Comment by nemanja-boric-sociomantic** [comment]
D1 helps you being more productive.

File: src/swarm/neo/client/mixins/RequestCore.d
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/RequestCore.d]: release notes?

Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/RequestCore.d]: Documentation also says "Private helper"

Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/RequestCore.d:313]: Documentation states "private helper", also perhaps missing release notes.

**Comment by gavin-norman-sociomantic** [comment]
Updated with release notes.

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/AllNodesRequestCore.d:152]: Would it be a problem if the `run()` exited because `this.connector()` failed, so the `this.initialiser.initialise()` was not yet executed?

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/AllNodesRequestCore.d]: This loop is not needed, indeed. The reason is that when `waitForReconnect` returns, we get either `connected = true`, or the program aborts (see `default` clause in switch).

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/AllNodesRequestCore.d]: typo `deinfed`

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/AllNodesRequestCore.d]: I almost said "there shouldn't be any sideeffects inside `assert`, since it would be removed when compiling with `-release`", but this is essentially a way to say `assert(this.conn.isConnected())`. Maybe we should make that clearer (not in this PR).

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/AllNodesRequestCore.d]: I wonder if we make all this field `private`, would that make compiler enforce that it's touched only by the other entities in this module. This would enforce that the user defined code doesn't change this field.

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/client/mixins/AllNodesRequestCore.d]: Nothing to do with the `all_nodes` in case of reset?

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (david-eckardt-sociomantic) [reply src/swarm/neo/client/mixins/AllNodesRequestCore.d]: :+1:
`waitForReconnect` seems to potentially have a side effect (that is, to wait), and IMHO assertions should not have any side effects regardless of whether code relies on them or not.

Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/AllNodesRequestCore.d]: Ah, you're right. In the case where the assertion will _fail_, there could be side-effects. That's obviously not what I intended. Yeah, `assert(this.conn.isConnected())` is what I intended. I'll have a look at how to get that information.

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/AllNodesRequestCore.d]: Right, it was just an old note to myself that I forgot to address before pushing.

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/AllNodesRequestCore.d:152]: I don't think so. `initialiser.reset` should just reset the initialiser to its initial, default state. If `initialiser.inisialise` hasn't been called, it should be in its default state anyway.

Did you have anything specific in mind?

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/AllNodesRequestCore.d]: Updated with an extra commit adding `client.RequestOnConn.AllNodesEventDispatcher.connection_status`.

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/AllNodesRequestCore.d]: In this PR, it can be. It's read externally in a subsequent PR, but I'll change it to private here and add a getter in the other PR.

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/AllNodesRequestCore.d]: Nope. `AllNodesRequestSharedWorkingData.initialising` is reset in the `scope(exit)` of `initialise` and `AllNodesRequestSharedWorkingData.called_started_notifier` should never be reset because it must be called strictly once ever during the lifetime of the request.

**Comment by gavin-norman-sociomantic** [comment]
Updated:
* Removed TODO / loop.
* Added connection status getter, use that in assert.
* Made `AllNodesRequestSharedWorkingData.initialising` private.
* Fixed typo.

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (gavin-norman-sociomantic) [reply src/swarm/neo/client/mixins/AllNodesRequestCore.d]: I've added this explanation to the method doc.

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/client/mixins/AllNodesRequestCore.d:152]: Nope, that's all clarification I need. Thanks!

File: src/swarm/neo/client/mixins/AllNodesRequestCore.d
Comment (nemanja-boric-sociomantic) [reply src/swarm/neo/client/mixins/AllNodesRequestCore.d]: :+1: I wasn't aware of the next PR.

**Review by nemanja-boric-sociomantic — APPROVED** [review]

**Pull request merged**
